### ISSUE TITLE
Another improvement to ctellraw performance?

### ISSUE THREAD

**Issue by bobhenl** [Zrips/CMI#5136]
Hi, I just wanna ask about if there couldn't be another improvement to ctellraw performance. I use this cmd very often in almost all messages...

> Reverting back change where broadcast and ctellraw was moved into async mode causing issue with commands being performed in a wrong order when we need to perform quite few of those in one go. Its still strongly recommended to use \n to separate different lines instead of re-firing same command which involves bunch of checks which can be avoided if we would do it only once instead of 10 times in a row.

For ex. this didn't happen to me before when you changed it to async, but if someone experienced it, then I understand. But may there could be another solution for making it less heavy on performance?

Or doyou suggest me using another cmd instead of this?

All I want is just send message that could be used by another plugins, I can't use "msg" because there'll be message that console sent this msg etc... I just wanna clear command to define like in ctellraw, but sometimes I don't even need hover etc

**Comment by Zrips** [comment]
Issue is that people like to list bunch of commands in a row to get multiple lines shown for the player. Which is reasonable expectation. Issue comes when you have 10 commands to show 10 lines and it needs to ne shown to everyone on the server and then each line containing placeholders needs to be updated and sent to every single player on server 10 times. While better solution would be to have one longer command which gets processed once and then sent to all players. 
Async causes issue with multiple commands as each command gets put into separate thread and at moment its done being processed its been sent to the player, which is not always a right order. So some more complex approach is needed in this case. I need to introduce some command filtering for specialized commands and processes some of the commands in async and group them while keeping others in sync. Not all commands can be performed in async, so this is kind of hard coding each command which can be performed and which cant, then some weird filter and grouping in case you have different types of commands and so on... Will revisit this command and check what i could do to optimize it without actually moving into async, i might get some tiny boost here.

**Comment by Zrips** [comment]
And by the way, you can always create ctext and show that one instead of performing ctellraw command multiple times. In any case did some work on it as managed to get some decent performance boost when it comes to simple messages without any dynamic variables. Will do some more investigation what can be optimized for more robust approach.

**Comment by bobhenl** [comment]
But when I use ctellraw to show just to one player, doesn't it have same effect as ctext? I use this very often:
![image](https://user-images.githubusercontent.com/18662086/109002067-ffe7e100-76a5-11eb-93db-5b2b4bac9426.png)
```
    - asConsole! cmi ctellraw [playername] <T>&c&lRádce &f» &fBylo ti aktivováno &emoderátorské
      létání &fpo dobu &e5 min&f. Kliknutím aktivuješ po vypršení.</T>
```
And so much more... Just these simple one.

So, should I still recreate these to ctexts? But I think it's not so good way to have bunch of ctexts just for this. Or will it have same effect in the latest update?

**Comment by Zrips** [comment]
Difference between ctellraw and ctext is that ctext is pregenerated while ctellraw needs to be properly processed before we got that same ctext which can be shown to the player.

**Comment by bobhenl** [comment]
So should I recreate every ctellraw into ctext? And use ctellraw just for announces?

**Comment by CloudeLecaw** [comment]
> So should I recreate every ctellraw into ctext? And use ctellraw just for announces?

You can use it everywhere when u have a couple of ctellraws in one command. If u just have one ctellraw in ycur command it does not worth it to create a extra ctext ;)

**Comment by bobhenl** [comment]
Mhm, but the problem is that I mainly use just one ctext, I don't use more of them in one command, only when I need to send different message to the one player, then to the second, then to all of them. But for ex. when player votes 5 times then is better to show ctext or ctellraw? It's one ctellraw in command, but when player joins server and it's executed 5 times, I'm not sure then. BUT this ctellraw is just for one player, not for all, so does it have so big impact on performance? And then, when players vote for ex. every 5 second on different votesite (max 5), then the mesage to all + the specific player is shown every 5s, here will be better to change it to ctext or does it have same impact? Cause I still can't show one ctext on all votes, Istill need to send it so many times after each vote.

But otherwise, if I use it just once in commands, shouldn't I still change it to ctext due to ctext pregeneration? I use this as mentioned for vote, then for playtimerewards and for bunch of messages, but I mostly show it just to 1 player, in some cases for more (like global message when opening crates, global message that someone voted etc, but still just 1 message, in vote is problem that this message could be sent multiple times depending on when players are voting after previous vote).

**Comment by Zrips** [comment]
Having something progenerated which is constantly used is always better, its not some major difference, but if you have same text repeating over and over again then its always better to optimize it.

**Comment by bobhenl** [comment]
OK, I pregenerated almost all commands, combined with msg, broadcast and so on. But may, the last question here probably, what about supporting transfer variables (arguments) to ctext? So when I have command where I use $1 $2 etc...., couldn't it be transfered to ctext to use it pregenerated to be more better for performance? I found workaround for it too (by using one broadcast/msg + second line with ctext), but this would be nice too maybe.
